### PR TITLE
Add derive module

### PR DESCRIPTION
**Pull request by mgrover1**
Add a separate derive module to deal with this logic

### CODE REVIEW

File: funnel/collection/derive_variables.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
+import warnings
+from typing import Callable
+
+import pydantic
+from toolz import curry
+
+from funnel.collection.registry import derived_variable_registry, query_dependent_operator_registry
```
Comment (andersy005) [inline funnel/collection/derive_variables.py:7]:
```suggestion
from .registry import derived_variable_registry, query_dependent_operator_registry
```

File: funnel/collection/derive_variables.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+
+@pydantic.dataclasses.dataclass
+class Derived_Variable(object):
```
Comment (andersy005) [inline funnel/collection/derive_variables.py:11]:
```suggestion
class Derived_Variable:
```

File: funnel/collection/derive_variables.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    """
+    Support computation of variables that depend on multiple variables,
+    i.e., "derived vars"
+    """
+
+    dependent_vars: list
```
Comment (andersy005) [inline funnel/collection/derive_variables.py:17]:
```suggestion
    dependent_vars: List[str]
```

File: funnel/collection/derive_variables.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        return self.func(ds, **kwargs)
+
+    def _ensure_variables(self, ds):
+        """ensure that required variables are present"""
+        missing_var = set(self.dependent_vars) - set(ds.variables)
+        if missing_var:
+            raise ValueError(f'Variables missing: {missing_var}')
+
+
+class Query_Dependent_Operator(object):
```
Comment (andersy005) [inline funnel/collection/derive_variables.py:32]:
```suggestion
class QueryDependentOperator(object):
```

File: funnel/collection/derive_variables.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+
+@curry
+def register_query_dependent_operator(func, query_keys):
+    """register a function for computing derived variables"""
+    func_hash = hash(func)
+    if func_hash in query_dependent_operator_registry:
+        warnings.warn(f'overwriting query dependent operator "{func.__name__}" definition')
+
+    query_dependent_operator_registry[func_hash] = Query_Dependent_Operator(
```
Comment (andersy005) [inline funnel/collection/derive_variables.py:67]:
```suggestion
    query_dependent_operator_registry[func_hash] = QueryDependentOperator(
```

File: funnel/collection/derive_variables.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
+import warnings
+from typing import Callable
```
Comment (andersy005) [inline funnel/collection/derive_variables.py:2]:
```suggestion
from typing import Callable, List
```

**Comment by andersy005** [comment]
@mgrover1, could you add a few more tests when you get a moment? It's my understanding we should be able to test this without needing a "real" catalog

**Comment by mgrover1** [comment]
Sure - I can add a couple for the `QueryDependentOperator`

**Comment by andersy005** [comment]
@mgrover1, `s3fs` should be added to the environment file to address the latest test failures

**Comment by mgrover1** [comment]
Thanks- I couldn't figure out why it was failing....

**Comment by andersy005** [comment]
> Thanks- I couldn't figure out why it was failing....

Of course...

File: requirements.txt [hunk 7]
```diff
@@ -8,3 +8,4 @@ fsspec
 joblib
 sqlalchemy-utils
 sqlalchemy>=1.4.21
+s3fs
```
Comment (andersy005) [inline requirements.txt:11]:
```suggestion
```

Let's add this to the `environment/*.yml` instead

**Comment by mgrover1** [comment]
@andersy005 is this good to go?

**Review by andersy005 — APPROVED** [review]
Looks good to me 👍🏽

**Pull request merged**